controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -0,0 +1,312 @@
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/blang/semver"
+	"github.com/coredns/corefile-migration/migration"
+	"github.com/docker/distribution/reference"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeadmv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/types/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	corefileKey       = "Corefile"
+	corefileBackupKey = "Corefile-backup"
+	coreDNSKey        = "coredns"
+)
+
+type coreDNSMigrator interface {
+	Migrate(currentVersion string, toVersion string, corefile string, deprecations bool) (string, error)
+}
+
+type CoreDNSMigrator struct{}
+
+func (c *CoreDNSMigrator) Migrate(fromCoreDNSVersion, toCoreDNSVersion, corefile string, deprecations bool) (string, error) {
+	return migration.Migrate(fromCoreDNSVersion, toCoreDNSVersion, corefile, deprecations)
+}
+
+type coreDNSInfo struct {
+	Corefile       string
+	Deployment     *appsv1.Deployment
+	CurrentVersion string
+}
+
+// UpdateCoreDNS updates the kubeadm configmap, coredns corefile and coredns
+// deployment.
+func (w *Workload) UpdateCoreDNS(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane) error {
+	info, err := w.getCoreDNSInfo(ctx)
+	if err != nil {
+		return err
+	}
+	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration == nil {
+		return nil
+	}
+
+	// Validate the image tag.
+	if err := validateCoreDNSImageTag(info.CurrentVersion, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag); err != nil {
+		return errors.Wrapf(err, "failed to validate CoreDNS")
+	}
+
+	// Get the DNS cluster configuration.
+	dns := &kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS
+
+	// Perform the upgrade.
+	if err := w.updateCoreDNSImageInfoInKubeadmConfigMap(ctx, dns); err != nil {
+		return err
+	}
+	if err := w.updateCoreDNSCorefile(ctx, info, dns); err != nil {
+		return err
+	}
+	if err := w.updateCoreDNSDeployment(ctx, info, dns); err != nil {
+		return errors.Wrap(err, "unable to update coredns deployment")
+	}
+	return nil
+}
+
+// validateCoreDNSImageTag returns error if the versions don't meet requirements.
+// Some of the checks come from
+// https://github.com/coredns/corefile-migration/blob/v1.0.6/migration/migrate.go#L414
+func validateCoreDNSImageTag(fromVersion, toVersion string) error {
+	from, err := semver.Parse(fromVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", fromVersion)
+	}
+	to, err := semver.Parse(toVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", toVersion)
+	}
+	if from.Compare(to) >= 0 {
+		return fmt.Errorf("toVersion %q must be greater than fromVersion %q", to.String(), from.String())
+	}
+
+	// check if the from version is even in the list of coredns versions
```

Comment:
If we still need this method (considering [this comment](https://github.com/kubernetes-sigs/cluster-api/pull/2574/files#r390062606)), then I'm wondering if this should also verify the to version is supported as well.

Reply:
@ncdc mentioned to move this in a webhook, can we file an issue now or is that release blocking?

Reply:
I think following up via issue is fine.

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/2620